### PR TITLE
BaoSwap contract

### PR DESCRIPTION
**Pull request by CrystalPony**
This is a swap contract between any two ERC-20 tokens (including ERC-677).

It implements the following functions:

- **deposit(token, amount)**, allows a user to deposit a custom amount of one of the supported tokens
- **withdraw(token)**, allows a user to withdraw either of the tokens after 1 block
- **setFee(fee)**, allows the admin to set the fee in % (default: 0)
- **transferOwnership(newOwner)**, allows the admin to transfer ownership to another account
- **withdrawOnlyAdmin(token, amount)**, allows the admin to withdraw any amount of any token
- **pause**, **unpause**, allows the admin to pause/unpause the contract

Deposit can only be called if the contract is not paused, if there is no pending deposit, if there is enough allowance, and if the amount is enough to perform a swap and deduct a fee.
Withdraw can only be called after the deposit block and will transfer the whole deposit to the caller, converting it to the desired token if needed.
Conversion respects decimals places, and SafeMath is used throughout the contract.
The fee gets deducted at deposit time.

The contract also implements the following useful view-only functions:

- **pendingBalance(account,token)**
- **canWithdraw(account)**
- **withdrawableBalance(account, token)**
- **swapWithFee(fromToken, amount)**, calculates the amount after swapping and applying fee

This PR includes unit-tests that cover all of the functionality and the important considerations:

swapping between two tokens of different decimal denominations, admin command restrictions, applying fees to decimal amounts etc.

I've deployed the contract to xDai where it can be tested live:

https://blockscout.com/poa/xdai/address/0x6860CBA30957b629B655D970728D06D4F1bE63C0/contracts

Here are swaps that I've performed myself:

**- 0% fee:**

**Deposit 20 mock -> Withdraw 20 mock**

https://blockscout.com/xdai/mainnet/tx/0xd58acc21b97a0dfaca11f987df2e6f0e09a23bd8dbd7e9757bcca34d3793ddd1/token-transfers
https://blockscout.com/xdai/mainnet/tx/0xb5731874fcb042ee5ec42d8d686cba40f40eeed152cac574ee356768fe40724e/token-transfers

**Deposit 20 mock -> Withdraw 20 BAO**

https://blockscout.com/xdai/mainnet/tx/0x176fad605d76e9990c652d25a07deb870dd9dbedbc10ce1a4b0cfeee15919926/token-transfers
https://blockscout.com/xdai/mainnet/tx/0xf291305a436fd89884b7a10abbca7cbb34b43851832ccf4c126e8ddf96728009/token-transfers

**Deposit 20 BAO -> Withdraw 20 mock**

https://blockscout.com/xdai/mainnet/tx/0xc80f641b86de7218e3c75f8a34dc311d613daa2d6b2f66f6d1709005c1f97fea/token-transfers
https://blockscout.com/xdai/mainnet/tx/0x03935c811cf4c1d1190c0385e2de3f01901ac4cbbf9676332fd065a319db9938/token-transfers

**- 7% fee:**

**Set fee to 7%**

https://blockscout.com/xdai/mainnet/tx/0x0cd67db06e4f732047d922ec4f389a7af08ecdbef9b1eb96235491a292c47dfc/internal-transactions

**Swap 20.12345678 mock for 18.7148148054 BAO**

image

https://blockscout.com/xdai/mainnet/tx/0x905829ef63bbf5ee43365f85752d294ef7c3bd01bfdd4d6e71a9f70314a24c94/token-transfers
https://blockscout.com/xdai/mainnet/tx/0xc39ce46b77a39b00ad16d6937cd1fcad8f467e9e88a78fb5c0922b4307dbdca0/token-transfers

**Swap 20.123456789012345678 BAO for 18.71481481 mock**

image

https://blockscout.com/xdai/mainnet/tx/0x70a0bbcc476bd1db86959ae3fcc3283e72489860e92faba5c726ae456a1ba7ff/token-transfers
https://blockscout.com/xdai/mainnet/tx/0x89b2ed58775a01f15636dcc39b85f0ab0c6e952930e101fc57e9903f81e41529/token-transfers

### CODE REVIEW

**Comment by CrystalPony** [comment]
When playing with the contract on xDai I noticed that the last digit sometimes seems to round up.
Turns out that the fee calculation was done in a way that caused loss of precision for the fee.

I fixed that in ce00ea7 and added a unittest for rounding-prone swaps.

File: README.md [hunk 1]
```diff
@@ -1 +1,9 @@
-# contracts
\ No newline at end of file
+# Contracts
+
+### BaoSwap
```
Comment (phillipleblanc) [inline README.md:3]: I wonder if this would be confused with https://www.baoswap.com/ ?

Comment (CrystalPony) [reply README.md:3]: Yes, we should rename it probably. BaoSwapper?
@thebaoman

**Comment by mastilver** [comment]
I could be wrong (maybe I misinterpreted the requirements) but I don't think it should be a swap as you implemented but rather a deposit contract where you can withdraw any tokens, see from #1 :

> The ability for users to deposit either token and have it added as a pending balance.
> The ability for users to withdraw their balance in their chosen version of the token.

I believe you should be able to deposit BAO.cx and withdraw BAO.cx

If I'm right, you should be able to remove `depositedToken` and combine the tokens balances into a single one

File: contracts/BaoSwap/BaoSwap.sol [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+	function canWithdraw(address account) public view returns (bool) {
+	    return _pendingBalances[account].pendingBlock <= block.number && _pendingBalances[account].balance1 > 0;
+	}
+	
+	function withdrawableBalance(address account, address token) onlySupportedToken(token) public view returns (uint256) {
+		if (_pendingBalances[account].pendingBlock > block.number) {
+			return 0;
+		}
+		if (token == _pendingBalances[account].depositedToken) {
+			return _pendingBalances[account].balance1;
```
Comment (mastilver) [inline contracts/BaoSwap/BaoSwap.sol:86]: you need to check the contract token balance, you can't withdraw if there is not enough tokens to do so

**Comment by CrystalPony** [comment]
> I could be wrong (maybe I misinterpreted the requirements) but I don't think it should be a swap as you implemented but rather a deposit contract where you can withdraw any tokens, see from #1 :
> 
> > The ability for users to deposit either token and have it added as a pending balance.
> > The ability for users to withdraw their balance in their chosen version of the token.
> 
> I believe you should be able to deposit BAO.cx and withdraw BAO.cx
> 
> If I'm right, you should be able to remove `depositedToken` and combine the tokens balances into a single one

This is implemented here - after depositing a token you can withdraw either the token you deposited or swap to the other one.

**Comment by mastilver** [comment]
right, my bad

**Comment by CrystalPony** [comment]
> If I'm right, you should be able to remove `depositedToken` and combine the tokens balances into a single one

Both balances are the same, the difference is the number of decimals.
My thinking was it's better to calculate this on deposit and not do decimal conversion every time withdrawableBalance is called.
depositedToken is used to choose between the balances. It could be done without it, but it's a cleaner implementation.
All this does store more data though, so it's probably worse from pure gas perspective - I'm not sure. Is this worth considering from gas perspective at all?

If this solution stays, internal balance1 and balance2 should probably be renamed not to be confused with token numeration.

**Pull request merged**